File: SortSnapshot/Classes/SortSnapshot+Sections.m
//
//  SortSnapshot+Sections.m
//  SortSnapshotDemoSwift
//
//  Created by Jinwoo Kim on 6/14/21.
//

#import "SortSnapshot+Sections.h"

@implementation NSDiffableDataSourceSnapshot (SortSnapshot_Sections)

- (void)ssSortSectionsUsingComparator:(NSComparator NS_NOESCAPE)cmptr {
    NSMutableArray *tmpSectionIdentifiers = [self.sectionIdentifiers mutableCopy];
    
    if (tmpSectionIdentifiers.count < 2) return;
    
    for (NSUInteger a = 0; a < tmpSectionIdentifiers.count; a++) {
        for (NSUInteger b = a + 1; b < tmpSectionIdentifiers.count; b++) {
            id aSectionIdentifier = tmpSectionIdentifiers[a];
            id bSectionIdentifier = tmpSectionIdentifiers[b];
            NSComparisonResult result = cmptr(aSectionIdentifier, bSectionIdentifier);
            BOOL shouldSwap;
            
            switch (result) {
                case NSOrderedAscending:
                    shouldSwap = NO;
                    break;
                case NSOrderedSame:
                    shouldSwap = NO;
                    break;
                case NSOrderedDescending:
                    shouldSwap = YES;
                    break;
                default:
                    shouldSwap = NO;
                    break;
            }
            
            if (shouldSwap) {
                id beforeBSectionIdentifier = tmpSectionIdentifiers[b - 1];
                
                [self moveSectionWithIdentifier:bSectionIdentifier beforeSectionWithIdentifier:aSectionIdentifier];
                
                if (![beforeBSectionIdentifier isEqual:aSectionIdentifier]) {
                    [self moveSectionWithIdentifier:aSectionIdentifier afterSectionWithIdentifier:beforeBSectionIdentifier];
                }
                
                [tmpSectionIdentifiers exchangeObjectAtIndex:a withObjectAtIndex:b];
            }
        }
    }
}

- (void)ssSortSectionsUsingDescriptors:(NSArray<NSSortDescriptor *> *)sortDescriptors {
    for (NSSortDescriptor *sortDescriptor in sortDescriptors) {
        NSComparator cmptr = ^NSComparisonResult(id obj1, id obj2){
            NSComparisonResult result = [sortDescriptor compareObject:obj1 toObject:obj2];
            return result * (sortDescriptor.ascending ? 1L : -1L);
        };
        
        [self ssSortSectionsUsingComparator:cmptr];
    }
}

@end

File: SortSnapshot/Classes/SortSnapshot+Sections.swift
//
//  SortSnapshot+Sections.swift
//  SortSnapshot
//
//  Created by Jinwoo Kim on 6/13/21.
//

#if os(iOS) || os(tvOS)
import UIKit
#else
import AppKit
#endif

extension NSDiffableDataSourceSnapshot {
    mutating func ssSortSections(by areInIncreasingOrder: (SectionIdentifierType, SectionIdentifierType) -> Bool) {
        var tmpSectionIdentifiers: [SectionIdentifierType] = sectionIdentifiers
        
        guard tmpSectionIdentifiers.count >= 2 else {
            return
        }
        
        for a in 0 ..< tmpSectionIdentifiers.count {
            for b in (a + 1) ..< tmpSectionIdentifiers.count {
                let aSectionIdentifier: SectionIdentifierType = tmpSectionIdentifiers[a]
                let bSectionIdentifier: SectionIdentifierType = tmpSectionIdentifiers[b]
                let shouldSwap: Bool = !(areInIncreasingOrder(aSectionIdentifier, bSectionIdentifier))
                
                if shouldSwap {
                    let beforeBSectionIdentifier: SectionIdentifierType = tmpSectionIdentifiers[b - 1]
                    
                    moveSection(bSectionIdentifier, beforeSection: aSectionIdentifier)
                    
                    if beforeBSectionIdentifier != aSectionIdentifier {
                        moveSection(aSectionIdentifier, afterSection: beforeBSectionIdentifier)
                    }
                    
                    tmpSectionIdentifiers.swapAt(a, b)
                }
            }
        }
    }
}

extension NSDiffableDataSourceSnapshot where SectionIdentifierType: Comparable {
    mutating func ssSortSections() {
        ssSortSections(by: <)
    }
}

#if IS_XCODE_13_OR_LATER
@available(iOS 15.0, macOS 12.0, tvOS 15.0, *)
extension NSDiffableDataSourceSnapshot {
    mutating func ssSortSections<T: SortComparator>(using comparator: T) where T.Compared == SectionIdentifierType {
        ssSortSections { first, second in
            let order: SortOrder = comparator.order
            let result: ComparisonResult = comparator.compare(first, second)

            switch (order, result) {
            case (.forward, .orderedAscending):
                return true
            case (.forward, .orderedDescending):
                return false
            case (.forward, .orderedSame):
                return false
            case (.reverse, .orderedAscending):
                return false
            case (.reverse, .orderedSame):
                return true
            case (.reverse, .orderedDescending):
                return false
            }
        }
    }
}
#endif
